include/dlaf/communication/executor.h

```diff
@@ -11,25 +11,141 @@
+/// hpx::mpi::experimental::detail::async()` call. Note that this is the main API difference with
+/// `dlaf::comm::async` where MPI_Comm is not part of the executor (i.e. pool_executor). The only other
+/// API difference is that `hpx::async` returns hpx::future<int> where `int` is the error code returned
+/// by the MPI call.
+///
+/// The internal mechanism for handling MPI requests is quite different. All requests are placed into a
+/// global array and polled for completion with `MPI_Testany()` on task completion/yielding for all tasks
+/// for which polling is enabled (note `enable_user_polling()`).
+///
+
+/// An executor for MPI calls.
+class executor {
+  MPI_Comm comm_;
+  hpx::threads::executors::pool_executor ex_;
+
+public:
+  // Associate the parallel_execution_tag executor tag type as a default with this executor.
+  using execution_category = hpx::parallel::execution::parallel_execution_tag;
+
+  inline executor(MPI_Comm comm) : comm_(comm), ex_("default", hpx::threads::thread_priority_high) {}
+
+  inline executor(std::string pool, MPI_Comm comm)
+      : comm_(comm), ex_(pool, hpx::threads::thread_priority_high) {}
+
+  constexpr bool operator==(executor const& rhs) const noexcept {
+    return comm_ == rhs.comm_ && ex_ == rhs.ex_;
   }
-  catch (...) {
-    return false;
+
+  constexpr bool operator!=(executor const& rhs) const noexcept {
+    return !(*this == rhs);
   }
+
+  constexpr executor const& context() const noexcept {
+    return *this;
+  }
+
+  Communicator comm() const noexcept {
+    return comm_;
+  }
+
+  /// The function only accepts non-blocking MPI routines. The MPI library needs
+  /// to support MPI_THREAD_MULTIPLE.
+  ///
+  /// Example usage:
+  ///
+  /// ```
+  /// ex.async_execute(MPI_Irecv, recv_ptr, num_recv_elems, MPI_DOUBLE, src_rank, tag);
+  /// ```
+  ///
+  /// compared to the usual MPI non-blocking call:
+  ///
+  /// ```
+  /// MPI_Irecv(recv_ptr, num_recv_elems, MPI_DOUBLE, src_rank, tag, comm, &request);
+  /// ```
+  ///
+  /// there are two differences
+  ///
+  /// 1) MPI_Comm (comm) doesn't need to be specified as it is passed by the executor
+  /// 2) MPI_Request (request) is omitted as it is handled internally in async()
+  ///
+  /// Note that if starting the communication depends on a previous task, it is often a good idea to call
+  /// the `get()` method on a returned future to avoid weird constructs like future<future<>>.
+  ///
+  /// Example:
+  ///
+  /// auto comm_fut = prev_fut.then([=] {
+  ///   return dlaf::comm::async(...).get();
+  /// });
+  ///
+  /// or don't return the future from `dlaf::comm::async()` and just do:
+  ///
+  /// auto comm_fut = prev_fut.then([=] {
+  ///   dlaf::comm::async(...);
+  /// });
+  ///
+  template <typename F, typename... Ts>
+  hpx::future<void> async_execute(F f, Ts... ts) noexcept {
+    // TODO: docs why this is done here instead of within the task
+    // TODO: support for MPI_THREAD_SERIALIZED
```

Comment:
> I'm wondering if the solution you have in mind for MPI_THREAD_SERIALIZED is to serialize the communications, so that MPI_THREAD_SERIALIZE would be enough. Which sounds like, more or less, what we currently do in the algorithm with the TaskChain

To support MPI_THREAD_SERIALIZED it is sufficient to make sure that MPI calls are not submitted at the same time. This can be done via a global `hpx::lcos::mutex`. This is lot less strict then imposing a certain order on tasks which is what `TaskChain` is doing.

This was discussed with John some time ago. The upstream HPX MPI executor doesn't currently support MPI_THREAD_SERIALIZED. It should be rather straightforward to provide support though. I will implement it here for now, and bug John next week to see if he wants to add it upstream.